File: backend/Models/Ship.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace Backend.Models
{
    /// <summary>
    /// Represents a cargo ship that transports containers
    /// </summary>
    public class Ship
    {
        /// <summary>
        /// Unique identifier for the ship
        /// </summary>
        [Key]
        public int ShipId { get; set; }
        
        /// <summary>
        /// Name of the ship
        /// </summary>
        [Required]
        [MaxLength(100)]
        public string Name { get; set; }
        
        /// <summary>
        /// IMO (International Maritime Organization) number
        /// </summary>
        [MaxLength(20)]
        public string ImoNumber { get; set; }
        
        /// <summary>
        /// Ship's flag state/country of registration
        /// </summary>
        [MaxLength(50)]
        public string Flag { get; set; }
        
        /// <summary>
        /// Type of ship (Container Ship, Bulk Carrier, Tanker, etc.)
        /// </summary>
        [MaxLength(50)]
        public string Type { get; set; }
        
        /// <summary>
        /// Container capacity in TEU (Twenty-foot Equivalent Units)
        /// </summary>
        [Range(1, int.MaxValue, ErrorMessage = "Capacity must be positive")]
        public int Capacity { get; set; }
        
        /// <summary>
        /// Current operational status of the ship (At Sea, Docked, Loading, Unloading, In Transit)
        /// </summary>
        [Required]
        public string Status { get; set; }
        
        /// <summary>
        /// Overall length of the ship in meters
        /// </summary>
        public decimal? Length { get; set; }
        
        /// <summary>
        /// Beam (width) of the ship in meters
        /// </summary>
        public decimal? Beam { get; set; }
        
        /// <summary>
        /// Draft (depth underwater) in meters
        /// </summary>
        public decimal? Draft { get; set; }
        
        /// <summary>
        /// Gross tonnage of the ship
        /// </summary>
        public decimal? GrossTonnage { get; set; }
        
        /// <summary>
        /// Year the ship was built
        /// </summary>
        public int? YearBuilt { get; set; }
        
        /// <summary>
        /// Current GPS coordinates (latitude,longitude)
        /// </summary>
        public string Coordinates { get; set; }
        
        /// <summary>
        /// Current speed in knots
        /// </summary>
        public decimal? Speed { get; set; }
        
        /// <summary>
        /// Current heading in degrees (0-360)
        /// </summary>
        public decimal? Heading { get; set; }
        
        /// <summary>
        /// Next port of call
        /// </summary>
        public string NextPort { get; set; }
        
        /// <summary>
        /// Estimated time of arrival at next port
        /// </summary>
        public DateTime? EstimatedArrival { get; set; }
        
        /// <summary>
        /// Foreign key to the current port (nullable)
        /// </summary>
        public int? CurrentPortId { get; set; }
        
        /// <summary>
        /// Timestamp when the ship record was created
        /// </summary>
        public DateTime CreatedAt { get; set; }
        
        /// <summary>
        /// Timestamp when the ship record was last updated
        /// </summary>
        public DateTime UpdatedAt { get; set; }
        
        // Navigation properties
        /// <summary>
        /// The current port where the ship is docked (if any)
        /// </summary>
        [ForeignKey("CurrentPortId")]
        public Port CurrentPort { get; set; }
        
        /// <summary>
        /// The containers currently on this ship
        /// </summary>
        public ICollection<Container> Containers { get; set; }
        
        /// <summary>
        /// The container loading operations for this ship
        /// </summary>
        public ICollection<ShipContainer> ShipContainers { get; set; }
        
        /// <summary>
        /// Berth assignments for this ship
        /// </summary>
        public ICollection<BerthAssignment> BerthAssignments { get; set; }
    }
}

File: frontend/src/types/ship.ts
/**
 * Enhanced Ship interface matching the updated backend ShipDto
 * Includes all rich fields from the backend Model
 */
export interface Ship {
  // Core identification
  shipId: number;
  id?: number; // Frontend compatibility
  name: string;
  
  // Ship specifications
  imoNumber?: string; // International Maritime Organization number
  flag?: string; // Flag state/country of registration
  type?: string; // Container Ship, Bulk Carrier, Tanker, etc.
  capacity?: number; // Container capacity in TEU
  
  // Physical dimensions
  length?: number; // Length in meters
  beam?: number; // Beam (width) in meters
  draft?: number; // Draft (depth below waterline) in meters
  grossTonnage?: number; // Gross tonnage
  yearBuilt?: number; // Year the ship was built
  
  // Operational status
  status: string; // At Sea, Docked, Loading, etc.
  
  // Location and movement
  coordinates?: string; // GPS coordinates (latitude, longitude)
  speed?: number; // Current speed in knots
  heading?: number; // Current heading/direction in degrees
  
  // Route information
  nextPort?: string; // Next port destination
  estimatedArrival?: string; // ETA at next port
  currentPortId?: number; // Current port if docked
  
  // Container information
  containerCount?: number; // Current number of containers
  
  // Legacy compatibility fields
  currentLocation?: string; // Maps to coordinates or current port
  arrivalTime?: string; // Maps to estimatedArrival
  departureTime?: string; // For scheduled departure
  createdAt?: string;
  updatedAt?: string;
}

export interface CreateShipRequest {
  name: string;
  status?: string;
  capacity?: number;
  type?: string;
  flag?: string;
  currentLocation?: string;
}

export interface UpdateShipRequest {
  name?: string;
  imoNumber?: string;
  flag?: string;
  type?: string;
  capacity?: number;
  status?: string;
  length?: number;
  beam?: number;
  draft?: number;
  grossTonnage?: number;
  yearBuilt?: number;
  coordinates?: string;
  speed?: number;
  heading?: number;
  nextPort?: string;
  estimatedArrival?: string;
  currentPortId?: number;
  
  // Legacy compatibility
  currentLocation?: string;
}

export interface ShipContainer {
  id: number;
  shipId: number;
  containerId: string;
  loadedAt: string;
  unloadedAt?: string;
  position?: string;
  status: string;
}

/**
 * Ship route information for optimization and tracking
 */
export interface ShipRoute {
  id: number;
  shipId: number;
  routeNumber?: string;
  originPortId: number;
  destinationPortId: number;
  scheduledDeparture: string;
  scheduledArrival: string;
  actualDeparture?: string;
  actualArrival?: string;
  routeStatus: 'Scheduled' | 'InProgress' | 'Completed' | 'Delayed';
  weatherDelay: number; // hours
  portDelay: number; // hours
  fuelConsumption?: number; // liters
  createdAt: string;
  updatedAt: string;
}

/**
 * Real-time ship tracking data
 */
export interface ShipRealtimeStatus {
  shipId: number;
  status: 'at-sea' | 'docked' | 'loading' | 'arriving' | 'departing';
  coordinates?: string;
  speed?: number;
  heading?: number;
  lastUpdated: string;
  estimatedArrival?: string;
}

/**
 * Ship statistics for dashboard display
 */
export interface ShipStats {
  totalShips: number;
  dockedShips: number;
  atSeaShips: number;
  averageSpeed: number;
  totalContainersTransported: number;
}

export interface BerthAssignment {
  id: number;
  shipId: number;
  berthId: number;
  assignedAt: string;
  expectedDeparture?: string;
  actualDeparture?: string;
  status: string;
}
